docs/design/tracking-branches.md

```diff
@@ -242,50 +278,68 @@ Note: desired behavior of `jj branch forget` is to
-* Pushing/exporting new branch, untracked remote exists
-  1. Exports `[local, remote] - [absent]`
+* Pushing new branch, untracked remote exists
+  1. Pushes `[local, remote] - [absent]`
      * Fails if `local` moved backwards or sideways
   2. Sets `remotes[remote].branches[name].state = tracking`
   3. `import_refs()` merges `[local, local] - [remote]` -> `local` (noop)
-* Pushing/exporting existing branch to tracking remote
-  1. Exports `[local, remote] - [remote]` -> `local`
+* Pushing existing branch to tracking remote
+  1. Pushes `[local, remote] - [remote]` -> `local`
      * Fails if `local` moved backwards or sideways, and if `remote` is out of
        sync
   2. `import_refs()` merges `[local, local] - [remote]` -> `local` (noop)
-* Pushing/exporting existing branch to untracked remote
+* Pushing existing branch to untracked remote
   * Same as "new branch"
-* Pushing/exporting deleted branch to tracking remote
-  1. Exports `[absent, remote] - [remote]` -> `absent`
+* Pushing deleted branch to tracking remote
+  1. Pushes `[absent, remote] - [remote]` -> `absent`
      * TODO: Fails if `remote` is out of sync?
   2. `import_refs()` merges `[absent, absent] - [remote]` -> `absent`
   3. Removes `remotes[remote].branches[name]` (`target` becomes `absent`)
-* Pushing/exporting deleted branch to untracked remote
+* Pushing deleted branch to untracked remote
   * Noop since `[absent, remote] - [absent]` -> `remote`
   * Perhaps, UI will report error
 * Pushing forgotten branch to untracked remote
   * Same as "deleted branch to untracked remote"
-* Exporting forgotten branch
-  1. Local branch change is noop since `[absent, absent] - [absent]` -> `absent`
-  2. Exports `forgotten` state to the backing Git repo:
-    `[absent, known_remote] - [known_remote]` -> `absent`
-    (This includes local branch in the pseudo `"git"` remote)
-  3. Removes `remotes[remote].branches[name]` (`target` becomes `absent`)
 * Pushing previously-forgotten branch to remote
   * Same as "new branch, untracked remote exists"
-  * The known `target` of forgotten remote branch is `absent`
+  * The `target` of forgotten remote branch is `absent`
+
+### export
+
+* Exporting new local branch, git branch doesn't exist
+  1. Sets `remotes["git"].branches[name].target = local`, `.state = tracking`
+  2. Exports `[local, absent] - [absent]` -> `local`
+* Exporting new local branch, git branch is out of sync
+  1. Exports `[local, git] - [absent]` -> fail
+* Exporting existing local branch, git branch is synced
+  1. Sets `remotes["git"].branches[name].target = local`
+  2. Exports `[local, git] - [git]` -> `local`
+* Exporting deleted local branch, git branch is synced
+  1. Removes `remotes["git"].branches[name]`
+  2. Exports `[absent, git] - [git]` -> `absent`
+* Exporting forgotten branches, git branches are synced
+  1. Exports `[absent, git] - [git]` -> `absent` for forgotten local/remote
+     branches
+
+### undo fetch
```

Comment:
How about undoing push? I think how I ideally would want it to behave is like this:

1. A branch initially points to commit A both locally and on some remote
2. Move a branch locally from commit A to commit B
3. Push the branch to the remote
4. Restore the repo to step 1 (so the local branch moves back to A)
5. Pushing the branch to the remote again now moves the branch back to commit A on the remote

I remember seeing someone else on Discord hoping it would behave that way (IIUC).

Is that how it will behave? The diagram you added says that remote branches get restored by `jj op restore`, which seems to mean that we would not detect that the branch had not moved (it points to A locally and on the remote). I didn't follow the whole thread about `jj op restore --what`, but I think I'm saying that I think it should not restore remote branches by default.

Reply:
> How about undoing push?

We'll need to use `jj undo --what repo` to not restore remote branches. I don't think both of `fetch && undo` and `push && undo` behaviors can be supported. Each remote ref could have `pushed`/`fetched` (or `undoable = false/true`) flag, but restoring across multiple push/fetch operations would get messy.

FWIW, at the step 5, I would want to push the old remote branch, not the local branch which might have moved since the last fetch.

Reply:
> I don't think both of `fetch && undo` and `push && undo` behaviors can be supported.

Hmm, good point. That's unfortunate. I agree that it seems better to make `fetch && undo` work well.

> FWIW, at the step 5, I would want to push the old remote branch, not the local branch which might have moved since the last fetch.

My use case was that I had made some change and pushed, but then realized some mistake and just wanted to go back to the previous state. I wish I could have done `jj op restore && jj git push` for that.

Reply:
Maybe we can have separate fetch-base and push-base refs (as we have `remotes["git"]` and `git_refs` for import/export)? push-base refs and `git_refs` won't be rolled back on undo as they aren't the states managed by jj.

* `fetch && undo && export` -> restores to the original state (except for the push-base)
* `fetch && undo && fetch` -> equivalent to single `fetch`
* `squash && push && restore @-- && push` -> pushes pre-squash state
  (could be `squash && push && undo && push-refs name@remote`)
* `fetch && undo && push` -> overwrites remote by local change

Reply:
My gut feeling is that that's more complicated than it should be, but I don't have any better suggestion. I think your current proposal seems like the best balance so far. We can adjust it later if it's clearer to us then. My impression is that the separate fetch-base and push-base refs is a new idea needs more consideration, but let me know if you think it's actually a solid proposal that we should go with.

Reply:
> My impression is that the separate fetch-base and push-base refs is a new idea needs more consideration

Yeah, it will add another layer invisible to users, and is definitely more complicated. I'm closing this as resolved. I'll revisit it if I find another undo-fetch/push problem.